File: backend/app/services/health_record_service.py
# app/services/health_record_service.py
from typing import List
from prisma import Prisma
from prisma.errors import RecordNotFoundError

from app.schemas.health_alert_schemas import HealthRecordSchema


class HealthRecordService:
    @staticmethod
    async def get_student_health_records(
        prisma: Prisma, student_id: int, skip: int = 0, limit: int = 10
    ) -> List[HealthRecordSchema]:
        """
        Retrieve student's health records

        Args:
            prisma (Prisma): Prisma client
            student_id (int): ID of the student
            skip (int): Number of records to skip
            limit (int): Maximum number of records to return

        Returns:
            List[HealthRecordSchema]: List of health records for the student
        """
        try:
            health_records = await prisma.healthrecord.find_many(
                where={"student_id": student_id},
                skip=skip,
                take=limit,
                order_by={"date": "desc"},  # Order by date
            )

            return [HealthRecordSchema.from_orm(record) for record in health_records]
        except RecordNotFoundError:
            raise ValueError("No health records found for this student")
        except Exception as e:
            raise ValueError(f"Error retrieving health records: {str(e)}")

    @staticmethod
    async def create_health_record(
        prisma: Prisma, student_id: int, record_data: HealthRecordSchema
    ) -> HealthRecordSchema:
        """
        Create a new health record

        Args:
            prisma (Prisma): Prisma client
            student_id (int): ID of the student
            record_data (HealthRecordSchema): Health record details

        Returns:
            HealthRecordSchema: Created health record
        """
        try:
            # Validate student exists
            student = await prisma.student.find_unique(where={"id": student_id})
            if not student:
                raise ValueError("Student not found")

            # Create health record
            new_record = await prisma.healthrecord.create(
                data={
                    "student_id": student_id,
                    "provider_id": record_data.provider_id,
                    "diagnosis": record_data.diagnosis,
                    "prescription": record_data.prescription,
                    "date": record_data.date,
                }
            )

            return HealthRecordSchema.from_orm(new_record)

        except Exception as e:
            raise ValueError(f"Error creating health record: {str(e)}")

File: backend/src/services/HealthRecordService.js
const { PrismaClient } = require("@prisma/client");
const MedicalDocument = require("../models/MedicalDocument");
const { uploadFile } = require("../utils/storage");
const { ApiError } = require("../utils/apiError");
const logger = require("../utils/logger");

class HealthRecordService {
  constructor() {
    this.prisma = new PrismaClient();
  }

  async createHealthRecord(providerId, studentId, recordData) {
    return this.prisma.healthRecord.create({
      ...recordData,
      providerId,
      studentId,
    });
  }

  async uploadMedicalDocument(file, recordId) {
    const { buffer, mimetype, originalname } = file;
    const fileUrl = await uploadFile(buffer, originalname, mimetype);

    return MedicalDocument.createDocument(recordId, {
      name: originalname,
      storagePath: fileUrl,
    });
  }
  async recentUploads(providerId) {
    return this.prisma.healthRecord.findMany({
      where: { providerId },
      orderBy: { createdAt: "desc" },
      take: 10,
      include: {
        documents: true,
        student: {
          include: {
            profile: {
              include: {
                user: true,
              },
            },
          },
        },
      },
    });
  }
}

module.exports = new HealthRecordService();
